# AIPI 590 (Deep Reinforecement Learning) HOMEWORK 2, Question 1

In [1]:
NAME = "Archit Kaila"

## Question 1: Implement double DQN

#### Hint: 
- SB3 only implements (valinia) DQN (the one discussed in the class). This task asks you to implement double DQN and compare the performance of DQN with the one of double DQN. double DQN is a minor modification of DQN. The only difference is that double DQN uses two Q networks to calculate the target labels. The reason that double DQN uses two Q networks is to reduce the over estimation of DQN.

- More details can be found in the slides, CS285, Berkeley Lecture 8
https://rail.eecs.berkeley.edu/deeprlcourse/static/slides/lec-8.pdf

- There are two ways to complete this homework. You can either complete the exercises in the starter code notebook, https://colab.research.google.com/github/Stable-Baselines-Team/rl-colab-notebooks/blob/sb3/dqn_sb3.ipynb, and submit the notebook. The exercises in this notebook will help you implement double DQN. Or you can directly modify the source code of DQN in SB3. If you decide to directly modify the source code of DQN in SB3 (in particular, modify the bootstrapped target), to submit, please create another Colab notebook and do something like !pip install git+”your SB3 github URL”, and run the training/testing experiments in this notebook.

#### Requirements:
- Compare double DQN with DQN


#### Submission:
- Create a new Colab notebook
- !pip install git+”your github URL”
- Training/testing experiments using the Colab notebook



### Installing modified stable baseline3 (with Double DQN implementation)
- personal git url: "https://github.com/architkaila/stable-baselines3"

In [2]:
## Use pip to install modified stable baseline3
!pip install command git+https://github.com/architkaila/stable-baselines3

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning https://github.com/architkaila/stable-baselines3 to /tmp/pip-req-build-l82jj6de
  Running command git clone -q https://github.com/architkaila/stable-baselines3 /tmp/pip-req-build-l82jj6de
     |████████████████████████████████| 1.5 MB 33.0 MB/s 
  Created wheel for stable-baselines3: filename=stable_baselines3-1.6.1a4-py3-none-any.whl size=168469 sha256=dadd19b77e55aed2426da127e5d662757c755dc54b7e45126af6986e1379c46c
  Stored in directory: /tmp/pip-ephem-wheel-cache-gyofiz2z/wheels/9a/f7/36/57a887a1a5d174fb384283c2e2b94bc08aae2cecdd4d77754e
  Created wheel for gym: filename=gym-0.21.0-py3-none-any.whl size=1616827 sha256=90d885d053e28767fe23bff5f40ca20974676a93a756f83bdff698f27fa18b82
  Stored in directory: /root/.cache/pip/wheels/76/ee/9c/36bfe3e079df99acf5ae57f4e3464ff2771b34447d6d2f2148
  Created wheel for command: filename=Command-0.1.0-py3-none-any.whl size=6413 sha256=1a9b

### Importing Libraries

In [3]:
## Library Imports
import gym
import numpy as np

from stable_baselines3 import DQN
from stable_baselines3.common.evaluation import evaluate_policy

### Creating GYM Environments For Training & Evaluation 

In [4]:
## Training environment of Cartpole
train_env = gym.make('CartPole-v1')

## Evalutation environment of Cartpole
evaluation_env = gym.make('CartPole-v1')

### Training "Vanilla DQN" 
- use default enable_double_dqn value = False

In [5]:
## Initialize a Vanilla DQN
dqn_vanilla = DQN('MlpPolicy', 'CartPole-v1', verbose=1)

# Random agent, before training
reward_mean, reward_std = evaluate_policy(dqn_vanilla, evaluation_env, n_eval_episodes=100, deterministic=True)
print(f'Random DQN Mean Reward: {np.round(reward_mean,2)} +/- {np.round(reward_std,2)}')


Using cuda device
Creating environment from the given name 'CartPole-v1'
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.


/usr/local/lib/python3.7/dist-packages/stable_baselines3/common/evaluation.py:69: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  UserWarning,


Random DQN Mean Reward: 23.24 +/- 9.88


In [6]:
## Training the Vanilla DQN for 150000 steps
dqn_vanilla.learn(total_timesteps=150000)

Streaming output truncated to the last 5000 lines.
| train/              |          |
|    learning_rate    | 0.0001   |
|    loss             | 0.00989  |
|    n_updates        | 3813     |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 9.64     |
|    ep_rew_mean      | 9.64     |
|    exploration_rate | 0.05     |
| time/               |          |
|    episodes         | 3612     |
|    fps              | 2310     |
|    time_elapsed     | 28       |
|    total_timesteps  | 65291    |
| train/              |          |
|    learning_rate    | 0.0001   |
|    loss             | 0.00463  |
|    n_updates        | 3822     |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 9.6      |
|    ep_rew_mean      | 9.6      |
|    exploration_rate | 0.05     |
| time/               |          |
|    episodes         | 3616     |
|   

In [7]:
## Evaluating the Vanilla DQN using 100 episodes (After Training)
reward_mean, reward_std = evaluate_policy(dqn_vanilla, evaluation_env, n_eval_episodes=100, deterministic=True)
print(f'Trained DQN (Vanilla) Mean Reward: {np.round(reward_mean,2)} +/- {np.round(reward_std,2)}')

Trained DQN (Vanilla) Mean Reward: 275.75 +/- 52.06


### Training "Double DQN" 
- set enable_double_dqn value = True

In [8]:
## Initialize a Double DQN
dqn_double = DQN('MlpPolicy', 'CartPole-v1', verbose=1, enable_double_dqn=True)

## Random Double DQN, before training
reward_mean, reward_std = evaluate_policy(dqn_double, evaluation_env, n_eval_episodes=100, deterministic=True)
print(f'Random DQN Mean Reward: {np.round(reward_mean,2)} +/- {np.round(reward_std,2)}')

Using cuda device
Creating environment from the given name 'CartPole-v1'
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
Random DQN Mean Reward: 9.3 +/- 0.81


In [9]:
## Training the Double DQN for 100000 steps
dqn_double.learn(total_timesteps=150000)

Streaming output truncated to the last 5000 lines.
| train/              |          |
|    learning_rate    | 0.0001   |
|    loss             | 0.000974 |
|    n_updates        | 2840     |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 11.5     |
|    ep_rew_mean      | 11.5     |
|    exploration_rate | 0.05     |
| time/               |          |
|    episodes         | 3188     |
|    fps              | 2634     |
|    time_elapsed     | 23       |
|    total_timesteps  | 61400    |
| train/              |          |
|    learning_rate    | 0.0001   |
|    loss             | 0.0208   |
|    n_updates        | 2849     |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 11.5     |
|    ep_rew_mean      | 11.5     |
|    exploration_rate | 0.05     |
| time/               |          |
|    episodes         | 3192     |
|   

In [10]:
## Evaluating the Double DQN using 100 episodes (After Training)
reward_mean, reward_std = evaluate_policy(dqn_double, evaluation_env, n_eval_episodes=100, deterministic=True)
print(f'Trained Double DQN Mean Reward: {np.round(reward_mean,2)} +/- {np.round(reward_std,2)}')

Trained Double DQN Mean Reward: 208.93 +/- 43.54


## Final Results (Post Running)
- Mean Reward (Random Agent) : 23.24   +/- 09.88
- Mean Reward (Vanilla DQN)  : 275.75  +/- 52.06
- Mean Reward (Double DQN)   : 208.00  +/- 43.54